File: leetcode/generate_parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
      vector<string> result;
      helper(result, "", 0, 0, n);
      return result;
    }
    void helper(vector<string>& result, string current, int l, int r, int n) {
      if (l == n && r == n) {
        result.push_back(current);
        return;
      }
      if (l > r) helper(result, current + ")", l, r + 1, n);
      if (l < n) helper(result, current + "(", l + 1, r, n);
    }
};

File: leetcode/generate_parentheses.py
def generateParenthesis(n):
    def helper(result, cur, left, right, n):
        if (left == right and left == n): 
            result.append(cur)
            return
        if (left > right): helper(result, cur + ')', left, right + 1, n)
        if (left < n): helper(result, cur + '(', left + 1, right, n)
    result = []
    helper(result, "", 0, 0, n)
    return result

print generateParenthesis(3)